lib/node_modules/@stdlib/repl/lib/syntax_highlighter.js

```diff
@@ -0,0 +1,154 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2024 The Stdlib Authors.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+/* eslint-disable no-underscore-dangle, no-restricted-syntax, no-invalid-this */
+
+'use strict';
+
+// MODULES //
+
+var readline = require( 'readline' );
+var logger = require( 'debug' );
+var setNonEnumerableReadOnly = require( '@stdlib/utils/define-nonenumerable-read-only-property' );
+var tokenizer = require( './tokenizer.js' );
+var THEMES = require( './themes.js' );
+var ANSI = require( './ansi_colors.js' );
+
+
+// VARIABLES //
+
+var debug = logger( 'repl:syntax-highlighter' );
+
+
+// MAIN //
+
+/**
+* Constructor for creating a syntax-highlighter.
+*
+* @private
+* @constructor
+* @param {repl} repl - REPL instance
+* @param {WritableStream} ostream - writable stream
+* @returns {SyntaxHighlighter} syntax-highlighter instance
+*/
+function SyntaxHighlighter( repl, ostream ) {
+	if ( !( this instanceof SyntaxHighlighter ) ) {
+		return new SyntaxHighlighter( repl, ostream );
+	}
+	debug( 'Creating a new syntax-highlighter' );
+
+	// Cache a reference to the provided REPL instance:
+	this._repl = repl;
+
+	// Cache a reference to readline interface:
+	this._rli = repl._rli;
+
+	// Cache a reference to the output writable stream:
+	this._ostream = ostream;
+
+	// Cache a reference to the available themes:
+	this._themes = THEMES;
+
+	// Initialize a buffer containing the current theme:
+	this._theme = 'default';
+
+	// Initialize a buffer containing the current line to validate line changes:
+	this._line = '';
+
+	return this;
+}
+
+/**
+* Highlights the input line.
+*
+* @private
+* @name _highlightLine
+* @memberof SyntaxHighlighter.prototype
+* @type {Function}
+* @param {str} line - input line
+* @param {Array} tokens - array of tokens in the line
+* @returns {str} highlighted line
+*/
+setNonEnumerableReadOnly( SyntaxHighlighter.prototype, '_highlightLine', function highlightLine( line, tokens ) {
+	var highlightedLine;
+	var resetCode = ANSI[ 'reset' ];
+	var colorCode;
+	var colors = this._themes[ this._theme ];
+	var offset;
+	var token;
+
+	highlightedLine = line.split( '' ); // split into array for easy insertions
+	offset = 0; // track number of ANSI codes inserted
+	for ( token of tokens ) {
+		colorCode = ANSI[ colors[ token.type ] ];
+
+		// Highlight token if it's color exists in the theme...
+		if ( colorCode ) {
+			highlightedLine.splice( token.start + offset, 0, colorCode ); // insert colorCode
```

Comment:
This may be an instance where a linked list would be preferable (e.g., see `@stdlib/utils/linked-list`), as insertion and deletion would then be O(1). Here, repeated splicing is likely to result in repeated memory reallocation.

Reply:
But wouldn't inserting at specific indices would take O(k) time as you need to traverse the linked list? We have direct access in arrays so it will take O(k) time as well (for shifting after insertion). just curious.

Reply:
Why would you need to traverse the linked list each time? I think it depends on whether your tokens are ordered. If they are, then you shouldn't need to repeatedly traverse the list, as you'd simply need to walk the list (`highlightedLine`) to the next token start index.

Regarding "direct access in arrays", yes and no. While `splice` mutates, this does not mean that the pointer to the underlying memory remains constant. After enough insertions, the JS engine will need to copy existing contents to a new memory block large enough to accommodate the existing elements plus the number of new elements. While this allows memory to remain contiguous, it does mean that periodically the JS engine will need to perform a `memcpy`. A linked list is not guaranteed to be contiguous, but it does avoid the periodic memory allocations.

Accordingly, I'd expect linked list insertion and deletion to be more performant, even if the underlying memory is fragmented.

Reply:
> Why would you need to traverse the linked list each time?

Yes, I see what you mean. You need to walk to the next linked list node vs `list[ 0 ]`. However, it's still not O(k) for the direct array access due to the memory reallocation discussed above when the number of inserted elements causes the array to exceed its memory bounds. Typically engines will implement a scheme (e.g., repeated doubling) to allow an array to grow for a while without needing reallocation, but we don't have any means for knowing what this scheme is. As such, from first principles, I'd still expect a linked list to be more efficient on average.

Reply:
I just re-constructed the highlighted string instead of mutating an existing array or using a linked list.
I used a linked list in `tokenizer.js` instead where I needed to insert elements at the beginning of the list and linked list felt way more efficient there..

Reply:
@kgryte OK so I did some more refactoring and now the purpose of the function feels clearer. What do you think?